### ISSUE TITLE
Suspiciously high CPU usage whenever a window refreshes quickly

### ISSUE THREAD

**Issue by HadrienG2** [linuxmint/cinnamon#8460]
```
 * Cinnamon version 3.8.9
 * Distribution - openSUSE Tumbleweed 20190312
 * Graphics hardware *and* driver used - Tested on both AMD RX 560 w/ amdgpu and Quadro K620 w/ nvidia binary driver.
 * 64 bit
 * .xsession-errors: https://github.com/linuxmint/cinnamon/files/2967295/default.xsession-errors.txt
 ```

**Issue**

On my systems, cinnamon uses little CPU when nothing is happening. However, any window with a high refresh rate (video in web browser, digital audio workstation...) leads the cinnamon and X processes to consume a suspicious lot of CPU, 20~30% of one CPU combined when a maximized window refreshes at 60hz on a 1440p screen. How that is spread between X and cinnamon depends on the task at hand and the GPU driver in use.

System-wide profiling with perf reveals that ~1/3 of cinnamon's time is spent processing events in glib (see also issue #7937 ) and ~2/3 is spent drawing stuff in clutter:

```
-   22,57%     0,00%  cinnamon         libglib-2.0.so.0.5800.3             [.] g_main_context_iterate.isra.26
   - 22,57% g_main_context_iterate.isra.26
      - 14,00% g_main_context_dispatch
         - 13,68% clutter_clock_dispatch
            - 13,61% _clutter_stage_do_update
               + 13,57% clutter_stage_cogl_redraw
      - 6,10% g_main_context_prepare
           2,99% g_source_iter_next
         - 1,09% gdk_event_source_prepare
            + 1,08% XPending
         - 0,79% message_queue_prepare
            + dbus_connection_get_dispatch_status
      - 2,38% g_main_context_check
         - 1,25% gdk_event_source_check
            + 1,24% XPending
           1,04% g_source_iter_next
```

Expanding the clutter rendering code reveals that it's recursing to absurd levels (there are at least 50 nested stack frames below clutter_clock_dispatch on a hot path inside libclutter). This suggests that the renderer is either trying to render something very complex, or implemented in a very inefficient way. Neither of these are acceptable for a window compositing system that must withstand updates at the native screen refresh rate.

**Steps to reproduce**

- Open a system monitor on an empty desktop. CPU usage of X and cinnamon should be at 1-2%.
- Open a window that refreshes quickly, such as a video player.
- Look at your system monitor again. Cinnamon is now consuming 10~20% of a CPU core.

**Expected behaviour**

Cinnamon's CPU consumption should not be significantly affected by a fast-updating window on the desktop, and remain down at a few %. After all, when a window updates, all the compositor should be doing is redrawing the corresponding region of the screen by copying a bitmap square in the right place, and on modern graphics hardware that's very cheap.

**Comment by jaszhix** [comment]
A lot of compositor code was changed in 4.0 and git master. Please upgrade and re-open if it still occurs.

**Comment by Philipp91** [comment]
I'm experiencing something similar (25% CPU load, i.e. 100% of one core), but _only_ when a non-empty Cinnamon panel exists on my second monitor (which has VSync enabled, unlike the other monitor that has my main Cinnamon panel). I can reproduce with video playback or scrolling a webpage or so (no matter on which monitor). When I remove the Cinnamon panel from the second monitor, everything is fine. I have detailed some findings in [this post](https://forums.linuxmint.com/viewtopic.php?p=2007375#p2007375).

I've tried disabling VSync in several places (Mint's system settings, NVidia's driver settings, and `__GL_SYNC_TO_VBLANK=0` in `/etc/environment`), but have had no success.

I'm using Cinnamon 4.8.6 on Mint 20.1 with a 5.4.0-72-lowlatency kernel and a GeForce GTX 745 card with the closed-source driver version 460.73.01.
